File: projects/game_level_design/maze.m
% Example 1: Basic network
gridSize = 100;
numSegments = 200;
segmentLength = 7;
turnProbability = 0.25;
corridorMap = createRandomCorridorNetwork(gridSize, numSegments, segmentLength, turnProbability);
figure;
imshow(corridorMap);
title('Random Corridor Network');


function corridorNetwork = createRandomCorridorNetwork(gridSize, numSegments, segmentLength, turnProbability)
% createRandomCorridorNetwork creates a random corridor network in a grid.
%
%   corridorNetwork = createRandomCorridorNetwork(gridSize, numSegments, segmentLength, turnProbability)
%
%   Inputs:
%     gridSize        - Size of the square grid (e.g., 100 for 100x100).
%     numSegments     - Total number of corridor segments to generate.
%     segmentLength   - Length of each corridor segment (number of steps).
%     turnProbability - Probability (0-1) of changing direction after a segment.
%                       A lower value means straighter corridors.
%
%   Output:
%     corridorNetwork - A gridSize x gridSize matrix where 1s represent
%                       corridors and 0s represent walls.

    if nargin < 4
        turnProbability = 0.3; % Default turn probability
    end
    if nargin < 3
        segmentLength = 5; % Default segment length
    end
    if nargin < 2
        numSegments = 200; % Default number of segments
    end
    if nargin < 1
        gridSize = 100; % Default grid size
    end

    % Initialize the grid with walls (0s)
    corridorNetwork = zeros(gridSize, gridSize);

    % Choose a random starting point
    currentX = randi(gridSize);
    currentY = randi(gridSize);
    corridorNetwork(currentY, currentX) = 1; % Mark as corridor

    % Define possible directions: [dy, dx]
    % N: [-1, 0], S: [1, 0], E: [0, 1], W: [0, -1]
    directions = [-1, 0;    % North
                   1, 0;     % South
                   0, 1;     % East
                   0, -1];   % West

    currentDirectionIdx = randi(4); % Start with a random direction

    for i = 1:numSegments
        % Decide whether to turn or continue in the same direction
        if rand() < turnProbability
            currentDirectionIdx = randi(4); % Choose a new random direction
        end
        currentDirection = directions(currentDirectionIdx, :);

        % Walk for segmentLength
        for j = 1:segmentLength
            newX = currentX + currentDirection(2);
            newY = currentY + currentDirection(1);

            % Check boundaries
            if newX >= 1 && newX <= gridSize && newY >= 1 && newY <= gridSize
                currentX = newX;
                currentY = newY;
                corridorNetwork(currentY, currentX) = 1; % Mark as corridor
            else
                % If we hit a boundary, choose a new direction immediately
                % and break from this segment walk.
                currentDirectionIdx = randi(4);
                break;
            end
        end
    end
end

File: projects/game_level_design/maze.py
import matplotlib.pyplot as plt
import numpy as np


def create_random_corridor_network(grid_size: int = 100, num_segments: int = 200, segment_length: int = 5, turn_probability: float = 0.3) -> np.ndarray:
    corridor_network = np.zeros((grid_size, grid_size), dtype=int)
    current_x = np.random.randint(0, grid_size)
    current_y = np.random.randint(0, grid_size)
    corridor_network[current_y, current_x] = 1
    directions = np.array([[-1, 0], [1, 0], [0, 1], [0, -1]])
    current_direction_idx = np.random.randint(0, 4)

    for _ in range(num_segments):
        if np.random.random() < turn_probability:
            current_direction_idx = np.random.randint(0, 4)
        current_direction = directions[current_direction_idx]
        for _ in range(segment_length):
            new_x = current_x + current_direction[1]
            new_y = current_y + current_direction[0]
            if 0 <= new_x < grid_size and 0 <= new_y < grid_size:
                current_x = new_x
                current_y = new_y
                corridor_network[current_y, current_x] = 1
            else:
                current_direction_idx = np.random.randint(0, 4)
                break
    return corridor_network


grid_size = 100
num_segments = 200
segment_length = 7
turn_probability = 0.25
corridor_map = create_random_corridor_network(grid_size, num_segments, segment_length, turn_probability)
plt.imshow(corridor_map, cmap="gray")
plt.title("Random Corridor Network")
plt.axis("off")
plt.show()
